central/metrics/custom/tracker/tracker_base.go

```diff
@@ -276,21 +278,57 @@ func (tracker *TrackerBase[Finding]) Gather(ctx context.Context) {
 }
 
 // getGatherer returns the existing or a new gatherer for the given userID.
-// When creating a new gatherer, it also registers all known metrics on the
-// gatherer registry.
+// The returned gatherer will be set to a running state for synchronization
+// purposes. When creating a new gatherer, it also registers all known metrics
+// on the gatherer registry.
+// Returns nil on error, or if the gatherer for this userID is still running.
 func (tracker *TrackerBase[Finding]) getGatherer(userID string, cfg *Configuration) *gatherer {
-	// TODO(PR #16176): limit the number of different tokens accessing the metrics?
+	defer tracker.cleanupInactiveGatherers()
 	var gr *gatherer
 	if g, ok := tracker.gatherers.Load(userID); !ok {
+		r := tracker.registryFactory(userID)
 		gr = &gatherer{
-			registry: tracker.registryFactory(userID),
+			registry: r,
 		}
+		gr.running.Store(true)
 		tracker.gatherers.Store(userID, gr)
 		for metricName := range cfg.metrics {
 			tracker.registerMetric(gr, cfg, metricName)
 		}
 	} else {
 		gr = g.(*gatherer)
+		// Return nil if this gatherer is still running.
+		// Otherwise mark it running.
+		if !gr.running.CompareAndSwap(false, true) {
+			return nil
+		}
 	}
 	return gr
 }
+
+// cleanupInactiveGatherers frees the registries for the userIDs, that haven't
+// shown up for inactiveGathererTTL.
+func (tracker *TrackerBase[Finding]) cleanupInactiveGatherers() {
+	tracker.cleanupWG.Add(1)
+	go func() {
+		defer tracker.cleanupWG.Done()
+		tracker.gatherers.Range(func(userID, gv any) bool {
+			g := gv.(*gatherer)
+			// Make it running to block normal gathering.
+			if !g.running.CompareAndSwap(false, true) {
+				return true
+			}
+			if time.Since(g.lastGather) >= inactiveGathererTTL &&
+				// Do not delete a just created gatherer in test.
+				// Not in test the lastGather should never be zero for a
```

Comment:
```suggestion
				// Note that in tests the lastGather should never be zero for a
```

Reply:
Well, that should actually read 'In production the lastGatherer...' which is opposed to run in test. I'll update the comment to make it less confusing.